### PR TITLE
Django 4

### PR DESCRIPTION
**Pull request by joanico**
- [ ] Upgraded Django 4.1
- [ ] Upgraded Wagtail 4

### CODE REVIEW

**Review by marobo — APPROVED** [review]

File: pykafe/contact/models.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from wagtail.core.fields import RichTextField, StreamField
 from wagtail.contrib.forms.models import AbstractEmailForm, AbstractFormField
 from modelcluster.fields import ParentalKey
-from wagtail.images.edit_handlers import ImageChooserPanel
+#from wagtail.images.edit_handlers import ImageChooserPanel
```
Comment (PeteCoward) [inline pykafe/contact/models.py:6]:
```suggestion
```

File: pykafe/home/models.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
 from django.db import models
 from wagtail.core.models import Page
 from wagtail.core.fields import RichTextField, StreamField
-from wagtail.admin.edit_handlers import FieldPanel, StreamFieldPanel
-from wagtail.images.edit_handlers import ImageChooserPanel
+from wagtail.admin.edit_handlers import FieldPanel
+#from wagtail.images.edit_handlers import ImageChooserPanel
```
Comment (PeteCoward) [inline pykafe/home/models.py:5]:
```suggestion
```

File: pykafe/pykafe/settings/local.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import os
+import django
+from django.utils.encoding import force_str
+django.utils.encoding.force_text = force_str
+
+from django.utils.encoding import smart_str
+django.utils.encoding.smart_text = smart_str
```
Comment (PeteCoward) [inline pykafe/pykafe/settings/local.py:7]: I would recommend removing these lines, and changing the code which fails to use the force_str and smart_str methods

File: pykafe/pykafe/urls.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 from django.conf import settings
-from django.urls import include, path
+from django.urls import include, path
```
Comment (PeteCoward) [inline pykafe/pykafe/urls.py:2]:
```suggestion
from django.urls import include, path
```

File: pykafe/requirements.txt [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
+django-rosetta==0.9.8
 django-tracking2==0.5.0
+django-widget-tweaks==1.4.3
 geoip2==2.9.0
-wagtailcodeblock==1.15.0.0
-django-rosetta==0.9.3
-django-leaflet==0.24.0
-django-active-users==0.2.2
-django-redis==4.10.0
+psycopg2==2.9.5
```
Comment (PeteCoward) [inline pykafe/requirements.txt:9]: remove or change to `-binary`

File: pykafe/search/middlewares.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class AjaxMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        def is_ajax(self):
```
Comment (PeteCoward) [inline pykafe/search/middlewares.py:6]: I would recommend checking where we use the is_ajax method, is it worth keeping?

Comment (PeteCoward) [reply pykafe/search/middlewares.py:6]: If in a dependency, upgrade that dependency

File: pykafe/analytic/wagtail_hooks.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 
 from tracking.models import Visitor
-from django.utils.translation import ugettext_lazy as _
+#from django.utils.translation import ugettext_lazy as _
```
Comment (PeteCoward) [inline pykafe/analytic/wagtail_hooks.py:3]:
```suggestion
```

File: pykafe/pykafe/settings/local.py (hunk 3 above)
Comment (joanico) [reply pykafe/pykafe/settings/local.py:7]: Upgraded :

django-active-users==0.4.0
django-tracking2==0.5.1

**Comment by marobo** [comment]
All changes already included in PR https://github.com/pykafe/pykafe/pull/117

**Pull request closed without merging**